{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39675\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "GOLR_URL = 'https://solr.monarchinitiative.org/solr/golr/select'\n",
    "SEARCH_URL = 'https://solr.monarchinitiative.org/solr/search/select'\n",
    "\n",
    "### \n",
    "# What are all the human genes with\n",
    "\n",
    "# Set of all human genes in monarch\n",
    "human_genes = set()\n",
    "\n",
    "# Set of all genes with ortholog-phenotype associations\n",
    "ortho_pheno_genes = set()\n",
    "\n",
    "# Set of all human genes with data (minus homology)\n",
    "human_genes_data = set()\n",
    "\n",
    "# Get all human genes\n",
    "# Schema yaml is here: \n",
    "# https://github.com/monarch-initiative/monarch-app/blob/master/conf/golr-views/monarch-search-config.yaml\n",
    "\n",
    "filters = [\n",
    "    'taxon:\"{0}\"'.format(\"NCBITaxon:9606\"),\n",
    "    'category:\"gene\"',\n",
    "]\n",
    "\n",
    "search_params = {\n",
    "    'wt': 'json',\n",
    "    'q': '*:*',\n",
    "    'fq': filters,\n",
    "    'fl': 'id',\n",
    "    'rows': 1000,\n",
    "    'start': 0\n",
    "}\n",
    "\n",
    "resultCount = search_params['rows']\n",
    "\n",
    "while search_params['start'] < resultCount:\n",
    "    solr_request = requests.get(SEARCH_URL, params=search_params)\n",
    "    response = solr_request.json()\n",
    "    resultCount = response['response']['numFound']\n",
    "    \n",
    "    for doc in response['response']['docs']:\n",
    "        human_genes.add(doc['id'])\n",
    "        \n",
    "    search_params['start'] += search_params['rows']\n",
    "    \n",
    "\n",
    "#print(human_genes[:][0:5])\n",
    "print(len(human_genes))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:__main__:Processed 0 documents\n",
      "INFO:__main__:Processed 10000 documents\n",
      "INFO:__main__:Processed 20000 documents\n",
      "INFO:__main__:Processed 30000 documents\n",
      "INFO:__main__:Processed 40000 documents\n",
      "INFO:__main__:Processed 50000 documents\n",
      "INFO:__main__:Processed 60000 documents\n",
      "INFO:__main__:Processed 70000 documents\n",
      "INFO:__main__:Processed 80000 documents\n",
      "INFO:__main__:Processed 90000 documents\n",
      "INFO:__main__:Processed 100000 documents\n",
      "INFO:__main__:Processed 110000 documents\n",
      "INFO:__main__:Processed 120000 documents\n",
      "INFO:__main__:Processed 130000 documents\n",
      "INFO:__main__:Processed 140000 documents\n",
      "INFO:__main__:Processed 150000 documents\n",
      "INFO:__main__:Processed 160000 documents\n",
      "INFO:__main__:Processed 170000 documents\n",
      "INFO:__main__:Processed 180000 documents\n",
      "INFO:__main__:Processed 190000 documents\n",
      "INFO:__main__:Processed 200000 documents\n",
      "INFO:__main__:Processed 210000 documents\n",
      "INFO:__main__:Processed 220000 documents\n",
      "INFO:__main__:Processed 230000 documents\n",
      "INFO:__main__:Processed 240000 documents\n",
      "INFO:__main__:Processed 250000 documents\n",
      "INFO:__main__:Processed 260000 documents\n",
      "INFO:__main__:Processed 270000 documents\n",
      "INFO:__main__:Processed 280000 documents\n",
      "INFO:__main__:Processed 290000 documents\n",
      "INFO:__main__:Processed 300000 documents\n",
      "INFO:__main__:Processed 310000 documents\n",
      "INFO:__main__:Processed 320000 documents\n",
      "INFO:__main__:Processed 330000 documents\n",
      "INFO:__main__:Processed 340000 documents\n",
      "INFO:__main__:Processed 350000 documents\n",
      "INFO:__main__:Processed 360000 documents\n",
      "INFO:__main__:Processed 370000 documents\n",
      "INFO:__main__:Processed 380000 documents\n",
      "INFO:__main__:Processed 390000 documents\n",
      "INFO:__main__:Processed 400000 documents\n",
      "INFO:__main__:Processed 410000 documents\n",
      "INFO:__main__:Processed 420000 documents\n",
      "INFO:__main__:Processed 430000 documents\n",
      "INFO:__main__:Processed 440000 documents\n",
      "INFO:__main__:Processed 450000 documents\n",
      "INFO:__main__:Processed 460000 documents\n",
      "INFO:__main__:Processed 470000 documents\n",
      "INFO:__main__:Processed 480000 documents\n",
      "INFO:__main__:Processed 490000 documents\n",
      "INFO:__main__:Processed 500000 documents\n",
      "INFO:__main__:Processed 510000 documents\n",
      "INFO:__main__:Processed 520000 documents\n",
      "INFO:__main__:Processed 530000 documents\n",
      "INFO:__main__:Processed 540000 documents\n",
      "INFO:__main__:Processed 550000 documents\n",
      "INFO:__main__:Processed 560000 documents\n",
      "INFO:__main__:Processed 570000 documents\n",
      "INFO:__main__:Processed 580000 documents\n",
      "INFO:__main__:Processed 590000 documents\n",
      "INFO:__main__:Processed 600000 documents\n",
      "INFO:__main__:Processed 610000 documents\n",
      "INFO:__main__:Processed 620000 documents\n",
      "INFO:__main__:Processed 630000 documents\n",
      "INFO:__main__:Processed 640000 documents\n",
      "INFO:__main__:Processed 650000 documents\n",
      "INFO:__main__:Processed 660000 documents\n",
      "INFO:__main__:Processed 670000 documents\n",
      "INFO:__main__:Processed 680000 documents\n",
      "INFO:__main__:Processed 690000 documents\n",
      "INFO:__main__:Processed 700000 documents\n",
      "INFO:__main__:Processed 710000 documents\n",
      "INFO:__main__:Processed 720000 documents\n",
      "INFO:__main__:Processed 730000 documents\n",
      "INFO:__main__:Processed 740000 documents\n",
      "INFO:__main__:Processed 750000 documents\n",
      "INFO:__main__:Processed 760000 documents\n",
      "INFO:__main__:Processed 770000 documents\n",
      "INFO:__main__:Processed 780000 documents\n",
      "INFO:__main__:Processed 790000 documents\n",
      "INFO:__main__:Processed 800000 documents\n",
      "INFO:__main__:Processed 810000 documents\n",
      "INFO:__main__:Processed 820000 documents\n",
      "INFO:__main__:Processed 830000 documents\n",
      "INFO:__main__:Processed 840000 documents\n",
      "INFO:__main__:Processed 850000 documents\n",
      "INFO:__main__:Processed 860000 documents\n",
      "INFO:__main__:Processed 870000 documents\n",
      "INFO:__main__:Processed 880000 documents\n",
      "INFO:__main__:Processed 890000 documents\n",
      "INFO:__main__:Processed 900000 documents\n",
      "INFO:__main__:Processed 910000 documents\n",
      "INFO:__main__:Processed 920000 documents\n",
      "INFO:__main__:Processed 930000 documents\n",
      "INFO:__main__:Processed 940000 documents\n",
      "INFO:__main__:Processed 950000 documents\n",
      "INFO:__main__:Processed 960000 documents\n",
      "INFO:__main__:Processed 970000 documents\n",
      "INFO:__main__:Processed 980000 documents\n",
      "INFO:__main__:Processed 990000 documents\n",
      "INFO:__main__:Processed 1000000 documents\n",
      "INFO:__main__:Processed 1010000 documents\n",
      "INFO:__main__:Processed 1020000 documents\n",
      "INFO:__main__:Processed 1030000 documents\n",
      "INFO:__main__:Processed 1040000 documents\n",
      "INFO:__main__:Processed 1050000 documents\n",
      "INFO:__main__:Processed 1060000 documents\n",
      "INFO:__main__:Processed 1070000 documents\n",
      "INFO:__main__:Processed 1080000 documents\n",
      "INFO:__main__:Processed 1090000 documents\n",
      "INFO:__main__:Processed 1100000 documents\n",
      "INFO:__main__:Processed 1110000 documents\n",
      "INFO:__main__:Processed 1120000 documents\n",
      "INFO:__main__:Processed 1130000 documents\n",
      "INFO:__main__:Processed 1140000 documents\n",
      "INFO:__main__:Processed 1150000 documents\n",
      "INFO:__main__:Processed 1160000 documents\n",
      "INFO:__main__:Processed 1170000 documents\n",
      "INFO:__main__:Processed 1180000 documents\n",
      "INFO:__main__:Processed 1190000 documents\n",
      "INFO:__main__:Processed 1200000 documents\n",
      "INFO:__main__:Processed 1210000 documents\n",
      "INFO:__main__:Processed 1220000 documents\n",
      "INFO:__main__:Processed 1230000 documents\n",
      "INFO:__main__:Processed 1240000 documents\n",
      "INFO:__main__:Processed 1250000 documents\n",
      "INFO:__main__:Processed 1260000 documents\n",
      "INFO:__main__:Processed 1270000 documents\n",
      "INFO:__main__:Processed 1280000 documents\n",
      "INFO:__main__:Processed 1290000 documents\n",
      "INFO:__main__:Processed 1300000 documents\n",
      "INFO:__main__:Processed 1310000 documents\n",
      "INFO:__main__:Processed 1320000 documents\n",
      "INFO:__main__:Processed 1330000 documents\n",
      "INFO:__main__:Processed 1340000 documents\n",
      "INFO:__main__:Processed 1350000 documents\n",
      "INFO:__main__:Processed 1360000 documents\n",
      "INFO:__main__:Processed 1370000 documents\n",
      "INFO:__main__:Processed 1380000 documents\n",
      "INFO:__main__:Processed 1390000 documents\n",
      "INFO:__main__:Processed 1400000 documents\n",
      "INFO:__main__:Processed 1410000 documents\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28613\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Get all human genes with data in golr\n",
    "# schema here: \n",
    "# https://github.com/monarch-initiative/monarch-app/blob/master/conf/golr-views/oban-config.yaml\n",
    "\n",
    "filters = [\n",
    "    'subject_taxon:\"{0}\"'.format(\"NCBITaxon:9606\"),\n",
    "    'subject_category:\"gene\"',\n",
    "    '-relation_closure:\"RO:HOM0000001\"'\n",
    "]\n",
    "\n",
    "golr_params = {\n",
    "    'wt': 'json',\n",
    "    'q': '*:*',\n",
    "    'fq': filters,\n",
    "    'fl': 'subject',\n",
    "    'rows': 1000,\n",
    "    'start': 0\n",
    "}\n",
    "\n",
    "resultCount = golr_params['rows']\n",
    "\n",
    "while golr_params['start'] < resultCount:\n",
    "    solr_request = requests.get(GOLR_URL, params=golr_params)\n",
    "    response = solr_request.json()\n",
    "    resultCount = response['response']['numFound']\n",
    "    \n",
    "    for doc in response['response']['docs']:\n",
    "        human_genes_data.add(doc['subject'])\n",
    "        \n",
    "    if golr_params['start'] % 100000 == 0:\n",
    "        logger.info(\"Processed {} documents\".format(golr_params['start']))\n",
    "        \n",
    "    golr_params['start'] += golr_params['rows']\n",
    "    \n",
    "\n",
    "#print(human_genes_data[:][0:5])\n",
    "print(len(human_genes_data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:__main__:Processed 0 documents\n",
      "INFO:__main__:Processed 100000 documents\n",
      "INFO:__main__:Processed 200000 documents\n",
      "INFO:__main__:Processed 300000 documents\n",
      "INFO:__main__:Processed 400000 documents\n",
      "INFO:__main__:Processed 500000 documents\n",
      "INFO:__main__:Processed 600000 documents\n",
      "INFO:__main__:Processed 700000 documents\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "209594\n"
     ]
    }
   ],
   "source": [
    "# Get all genes with ortholog phenotype associations\n",
    "\n",
    "filters = [\n",
    "    'subject_category:\"gene\"',\n",
    "    'object_category:\"phenotype\"'\n",
    "]\n",
    "\n",
    "golr_params = {\n",
    "    'wt': 'json',\n",
    "    'q': '*:*',\n",
    "    'fq': filters,\n",
    "    'fl': 'subject_ortholog_closure',\n",
    "    'rows': 1000,\n",
    "    'start': 0\n",
    "}\n",
    "\n",
    "resultCount = golr_params['rows']\n",
    "\n",
    "while golr_params['start'] < resultCount:\n",
    "    solr_request = requests.get(GOLR_URL, params=golr_params)\n",
    "    response = solr_request.json()\n",
    "    resultCount = response['response']['numFound']\n",
    "    \n",
    "    for doc in response['response']['docs']:\n",
    "        if 'subject_ortholog_closure' in doc:\n",
    "            for gene in doc['subject_ortholog_closure']:\n",
    "                ortho_pheno_genes.add(gene)\n",
    "        \n",
    "    if golr_params['start'] % 100000 == 0:\n",
    "        logger.info(\"Processed {} documents\".format(golr_params['start']))\n",
    "        \n",
    "    golr_params['start'] += golr_params['rows']\n",
    "    \n",
    "\n",
    "#print(ortho_pheno_genes[:][0:5])\n",
    "print(len(ortho_pheno_genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28653\n"
     ]
    }
   ],
   "source": [
    "### We want the intersection of all human_genes and ortho_pheno_genes\n",
    "### Union of the above and human_genes_data\n",
    "\n",
    "human_with_op = ortho_pheno_genes.intersection(human_genes)\n",
    "\n",
    "human_genes_w_data = human_genes_data.union(human_with_op)\n",
    "\n",
    "\n",
    "print(len(human_genes_w_data))\n",
    "\n",
    "file = open(\"/home/kshefchek/gene_list.txt\", 'w')\n",
    "\n",
    "for gene in human_genes_w_data:\n",
    "    file.write(\"{}\\n\".format(gene))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "grep -P '^NCBIGene' gene_list.txt >genes_only_entrez.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
